src/pages/Environments.js

```diff
@@ -116,8 +120,165 @@ const DeleteAppModal = ({ app: { googleProject, appName, diskName, appType }, on
   ])
 }
 
+const MigratePersistentDisksBanner = ({ count }) => {
+  const deadline = new Date('01 January 2023 00:00 UTC')
+  return count > 0 && div({
+    style: {
+      position: 'absolute', top: topBarHeight, left: '50%', transform: 'translate(-50%, -50%)',
+      backgroundColor: colors.warning(0.15),
+      border: '2px solid', borderRadius: '12px', borderColor: colors.warning(),
+      zIndex: 8000 // Must be under 9000 to not draw over notifications
+    }
+  }, [
+    div({ style: { display: 'flex', alignItems: 'center', margin: '0.75rem 1.5rem 0.75rem 1.5rem' } }, [
+      icon('warning-standard', { size: 32, style: { color: colors.warning(), marginRight: '0.25rem' } }),
+      div([
+        strong([`You have ${Math.floor((deadline - Date.now()) / 86400000)} days to migrate ${count} shared persistent ${count > 1 ?
+          'disks' :
+          'disk'}. `]),
+        `Un-migrated disks will be DELETED after ${Utils.makeCompleteDate(deadline)}.`
+      ])
+    ])
+  ])
+}
+
+const MigratePersistentDiskCell = ({ disk, setMigrateDisk }) => div({
+  style: {
+    display: 'flex', flex: 1, flexDirection: 'column', textAlign: 'center',
+    height: '100%', margin: '-1rem', padding: '0.5rem 0 0 0',
+    backgroundColor: colors.danger(0.15), color: colors.danger()
+  }
+}, [
+  h(TooltipTrigger, { content: `This disk is shared between workspaces, which is no longer supported. Click "Migrate" to make copies for relevant workspaces.` }, [
+    div(['Offline', icon('warning-standard', { style: { marginLeft: '0.25rem', color: colors.danger() } })])
+  ]),
+  h(Link, { onClick: () => setMigrateDisk(disk), style: { wordBreak: 'break-word' } },
+    ['Migrate']
+  )
+])
+
+const MigratePersistentDiskModal = ({ disk, workspaces, onSuccess, onDismiss, contactSupport, deleteDiskId }) => {
+  const [migrating, setMigrating] = useState()
+  const [isSelected, setIsSelected] = useState({})
+  const [deleteDisk, setDeleteDisk] = useState(false)
+
+  const copyDiskToWorkspace =
+    ({ googleProject, namespace: saturnWorkspaceNamespace, name: saturnWorkspaceName }) => {
+      return reportErrorAndRethrow(`Error copying persistent disk to workspace '${saturnWorkspaceName}'`,
+        () => Ajax().Disks.disk(googleProject, disk.name).create({
+          ..._.pick(['size', 'blockSize', 'zone'], disk),
+          diskType: disk.diskType.label,
+          labels: { saturnWorkspaceNamespace, saturnWorkspaceName },
+          sourceDisk: _.pick(['googleProject', 'name'], disk)
+        })
+      )()
+    }
+
+  const migrateDisk = _.flow(onError(onDismiss), Utils.withBusyState(setMigrating))(async () => {
```

Comment:
The combination of `reportErrorAndRethrow` and `onError(onDismiss)` looks similar to `withErrorReportingInModal`.

https://github.com/DataBiosphere/terra-ui/blob/4c71dba6556eeccb520dc2038202529f4fa0fae2/src/libs/error.js#L43-L57

Reply:
Ooh, I missed that detail as well.

Reply:
Instead of showing one notification for each failure, you could collect all failed disks and show a notification with the list.

Reply:
Do we have code to do this already? Otherwise I'll make a follow up change.

Reply:
ImportData has something similar to what I had in mind. Use [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) to wait for all requests, then throw a custom error if any failed. And the usual `withErrorReporting` wrapper could be used with that.

https://github.com/DataBiosphere/terra-ui/blob/ff9c7a2b69eb4cc5f0be9d4b3d2956c4ab8ff076/src/pages/ImportData.js#L152-L171

Reply:
I'd rather not have to add a new component to layout these errors and I'm not expecting users to select many workspaces to copy to

Reply:
Right, but it'll be one error containing multiple messages - ie
Failed to copy disk to workspace A because ...
Failed to copy disk to workspace D because ...